File: 2024/3/main.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

func readData() ([]string, error) {
	data, err := os.ReadFile(filepath.Join(".", "data.txt"))
	if err != nil {
		return nil, err
	}
	return strings.Split(string(data), "\n"), nil
}

func formatData(rows []string) []string {
	return rows
}

func part1(data []string) int {
	multiplierRegex := regexp.MustCompile(`mul\(\d{1,3},\d{1,3}\)`)
	mulRegex := regexp.MustCompile(`(\d)+`)

	multipliers := multiplierRegex.FindAllString(strings.Join(data, ""), -1)

	sum := 0

	for _, mul := range multipliers {
		nums := mulRegex.FindAllString(mul, -1)
		a, err1 := strconv.Atoi(nums[0])
		b, err2 := strconv.Atoi(nums[1])
		if err1 == nil && err2 == nil {
			sum += a * b
		}

	}
	fmt.Println("sum", sum)
	return sum
}

func part2(data []string) int {
	multiplierRegex := regexp.MustCompile(`(do(n't)?\(\))|(mul\(\d{1,3},\d{1,3}\))`)
	mulRegex := regexp.MustCompile(`\d+`)
	instructionRegex := regexp.MustCompile(`do(n't)?\(\)`)
	sum := 0

	instruction := "do()"

	var multipliers []string
	for _, line := range data {
		matches := multiplierRegex.FindAllString(line, -1)
		multipliers = append(multipliers, matches...)
	}

	for _, mul := range multipliers {
		if instructionRegex.MatchString(mul) {
			instruction = mul
		} else {
			if instruction != "do()" {
				continue
			}
			nums := mulRegex.FindAllString(mul, -1)
			a, err1 := strconv.Atoi(nums[0])
			b, err2 := strconv.Atoi(nums[1])
			if err1 == nil && err2 == nil {
				sum += a * b
			}
		}
	}
	fmt.Println("sum", sum)
	return sum
}

func main() {
	data, err := readData()
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	formattedData := formatData(data)
	part1(formattedData)
	part2(formattedData)
}

File: 2024/3/main.rs
use std::fs;
use std::path::Path;
use std::error::Error;
use tokio;
use regex::Regex;

async fn read_data() -> Result<Vec<String>, Box<dyn Error>> {
    let contents = fs::read_to_string(Path::new("data.txt"))?;
    Ok(contents.lines().map(String::from).collect())
}

async fn format_data(rows: Vec<String>) -> Vec<String> {
		rows
}

fn part1(data: &[String]) -> i32 {
	let mut sum = 0;
    let re = Regex::new(r"mul\((\d{1,3}),(\d{1,3})\)").unwrap();
    
    for line in data {
        for cap in re.captures_iter(line) {
            let a: i32 = cap[1].parse().unwrap();
            let b: i32 = cap[2].parse().unwrap();
            sum += a * b;
        }
    }
    
    println!("sum: {}", sum);
    sum
}

fn part2(data: &[String]) -> i32 {
	let mut sum = 0;
	let re_all = Regex::new(r"(do(n't)?\(\))|(mul\(\d{1,3},\d{1,3}\))").unwrap();
	let re_nums = Regex::new(r"\d+").unwrap();
	
	let mut instruction = String::from("do()");
	
	let multipliers: Vec<String> = data.iter()
			.flat_map(|x| re_all.find_iter(x))
			.map(|m| m.as_str().to_string())
			.collect();
	
	for mul in multipliers {
			if mul.contains("do") {
					instruction = mul.clone();
			} else {
					if instruction != "do()" {
							continue;
					}
					let nums: Vec<i32> = re_nums.find_iter(&mul)
							.map(|m| m.as_str().parse::<i32>().unwrap())
							.collect();
					sum += nums[0] * nums[1];
			}
	}
	
	println!("sum: {}", sum);
	sum
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
	let data = read_data().await?;
	let formatted_data = format_data(data).await;

  part1(&formatted_data);
	part2(&formatted_data);
	
	Ok(())
}

